### ISSUE TITLE
Resuming with fault-tolerant training fails when DataLoader has num_workers>1

### ISSUE THREAD

**Issue by eladsegal** [Lightning-AI/lightning#12285]
## 🐛 Bug

Resuming with fault-tolerant training fails when DataLoader has `num_workers`>1:
```
Traceback (most recent call last):
  ...
  File "~/pytorch-lightning/pytorch_lightning/utilities/auto_restart.py", line 533, in _reload_dataloader_state_dict_automatic_map_dataset
    dataloader.dataset.load_state_dict(
  File "~/pytorch-lightning/pytorch_lightning/utilities/auto_restart.py", line 261, in load_state_dict
    self._cached_state_dict = _rotate_worker_indices(deepcopy(state_dict), latest_worker_id, num_workers)
  File "~/pytorch-lightning/pytorch_lightning/utilities/auto_restart.py", line 615, in _rotate_worker_indices
    raise MisconfigurationException("The `state` should contain `num_workers - 1` values.")
pytorch_lightning.utilities.exceptions.MisconfigurationException: The `state` should contain `num_workers - 1` values.
```

### To Reproduce
Change 
https://github.com/PyTorchLightning/pytorch-lightning/blob/d31126c3310af62d76ffb87368ae1e4cc5bbac93/pl_examples/fault_tolerant/automatic.py#L96-L97
to
```
return DataLoader(RandomGetItemDataset(3, 1), num_workers=2)
```

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): master ([d31126c](https://github.com/PyTorchLightning/pytorch-lightning/commit/d31126c3310af62d76ffb87368ae1e4cc5bbac93))

cc @carmocca @justusschock @awaelchli @ninginthecloud @akihironitta

**Comment by SpontaneousDuck** [comment]
+1 to this issue

**Comment by jhoareau** [comment]
This is also happening to us under 1.7.4.

**Comment by awaelchli** [comment]
Just for documentation, this ticket is about resolving essentially the fixme/failing test case here:
https://github.com/Lightning-AI/lightning/blob/f3d4d83462fccf83d953f3a8f6bdf32ec8b7bcd9/tests/tests_pytorch/utilities/test_auto_restart.py#L1281-L1284

which reproduces the error seen here.

**Comment by jhoareau** [comment]
I think this can be closed - fault-tolerant support was deleted from PL with 2.0

**Comment by awaelchli** [comment]
@jhoareau In principle yes. We haven't figured out how to address the problems mentioned here, and it can't be fixed in 2.0 because as you say, we removed the code.

This experimental feature for fault tolerance became more complex than we thought. IMO we took it to the extreme, and probably one step too far, with the capture of random state in the dataloader workers. It is difficult to fix it because a fix would basically require a redesign. 

However note, we do have the commitment to support 1.9.x long-term, and so bug fixes can also happen for that version independently. If we find that more users depend on this fix, we could bump the priority. But I think the chances are low that this can be completed due to its complexity. Let me know what you think.

**Comment by carmocca** [comment]
I created an issue proposing a more straightforward mechanism to implement this again at https://github.com/Lightning-AI/lightning/issues/17105. In case you want to track it, leave your thoughts, or support it.

I propose closing this issue as this was an experimental feature, we would only merge this fix if a community contributor was to write it.
